File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.java
class Solution {
    public boolean canPartition(int[] nums) {
        int n=nums.length;

        int sum=Arrays.stream(nums).sum();
        if(sum % 2 != 0 || n==1) return false;
        sum /= 2; 

        boolean dp[]=new boolean[sum+1];
        dp[0]=true;

        if(nums[0] <= sum)
        dp[nums[0]] = true;

        for(int j=1;j<n;j++)
        {
            boolean cur[]=new boolean[sum+1];
            cur[0]=true;
            for(int i=1; i<=sum; i++)
            {
                boolean notTaken = dp[i];
                boolean taken= false;

                if(nums[j] <= i)
                taken=dp[i-nums[j]];

                cur[i] = notTaken || taken;
            }
            dp = cur;
        }
    
        return dp[sum];
    }
}

File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.cpp
class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int sum = accumulate(nums.begin(),nums.end(),0);
        if(sum % 2 == 1) return false;
        int target = sum / 2;
        vector<bool>dp(target+1,false);
        int n = nums.size();
        dp[0] = true;
        for(int i = 0; i < n; i++){
            for(int x = target; x >= nums[i]; x--){
                dp[x] = dp[x] | dp[x - nums[i]];
            }
        }
        
        return dp[target];
    }
};